src/hyp_map.rs

```diff
@@ -70,18 +72,122 @@ impl SharedRegion {
             let vaddr = PageAddr::new(RawAddr::supervisor_virt(r.base().bits())).unwrap();
             let page_count = PageSize::num_4k_pages(r.size()) as usize;
             let pte_fields = PteFieldBits::leaf_with_perms(pte_perms);
+            let variable_mapped = matches!(
+                r.region_type(),
+                HwMemRegionType::Reserved(HwReservedMemType::HypervisorHeap)
+                    | HwMemRegionType::Reserved(HwReservedMemType::HypervisorPerCpu)
+            );
             Some(Self {
                 vaddr,
                 paddr,
                 page_count,
                 pte_fields,
+                variable_mapped,
+                page_size: None,
             })
         } else {
             None
         }
     }
 
-    // Map this region into a page table.
+    /// partition a HypMapRegion to use minimal pages
```

Comment:
I'm slow and having trouble parsing this function. Could you provide an overview of what's going on? Is it finding the biggest page size that fits, using that until exhausted, then using smaller ones?